### PR TITLE
QDB-2547 - Add client-api logging ability to Python API

### PR DESCRIPTION
**Pull request by solatis**
Adds logging capabilities to the Python API. Most interesting to review is mostly the logger.cpp / logger.hpp.

### CODE REVIEW

File: quasardb/pandas/__init__.py [hunk 1]
```diff
@@ -253,11 +275,18 @@ def write_dataframe(df, cluster, table, create=False, _async=False, chunk_size=5
                     fn = write_with[ct]
                     fn(i, v)
 
+        start = time.time()
+
+        logger.debug("push chunk of %d rows, fast?=%s, async?=%s", len(current_df.index), fast, _async)
+
+        if fast is True:
+            batch.push_async()
```
Comment (mkurdej) [inline quasardb/pandas/__init__.py]: Didn't you mean:
```suggestion
            batch.push_fast()
```
?

File: quasardb/pandas/__init__.py [hunk 2]
```diff
@@ -175,6 +186,7 @@ def read_dataframe(table, row_index=False, columns=None, ranges=None):
         columns = list(c.name for c in table.list_columns())
 
     if not row_index:
+        logger.debug("reading DataFrame from %d series", len(columns))
```
Comment (mkurdej) [inline quasardb/pandas/__init__.py]: series? or columns?

File: quasardb/logger.hpp [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+   */
+  logger(const std::string & module_name)
+  {
+
+    // TODO(leon): is there a way we can (safely) cache these calls? they should
+    //             remain the same for every python VM instance... is it safe to cache?
+    //
+    // NOTE(leon): looking at the pybind11 code, it appears they never cache anything.. so
+    //             that might hint that it's complex to get right
+    py::module logging = py::module::import("logging");
```
Comment (mkurdej) [inline quasardb/logger.hpp]: Wouldn't it work to make them, hmmm, function static variables?
```suggestion
    static py::module logging = py::module::import("logging");
```

File: quasardb/error.hpp [hunk 4]
```diff
@@ -140,6 +141,10 @@ void qdb_throw_if_error(qdb_error_t err, PreThrowFtor && pre_throw = detail::no_
         throw qdb::exception{err};
       };
     }
+
+    // HACKS(leon): we need to flush our log buffer a lot, ideally after every native qdb
+    //              call. Guess which function is invoked exactly at those moments?
+    qdb::native::flush();
```
Comment (mkurdej) [inline quasardb/error.hpp]: Why don't you flush when there is an error? I'd put it before the `if`. Or am I missing something?

File: quasardb/logger.cpp [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    qdb::native::_do_flush();
+  }
+}
+
+
+void
+qdb::native::_do_flush() {
+  // NOTE: this assumes a lock to _buffer-lock has been acquired, never call this
+  //       function directly!
+
```
Comment (mkurdej) [inline quasardb/logger.cpp]: `assert(!_buffer_lock.try_lock());`
hmm?

File: quasardb/logger.hpp [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+   * As such, try to keep this logging out of the critical performance path.
+   */
+class logger
+{
+public:
+
+  /**
+   * Default constructor, invoking any .debug() etc on a default initialized
+   * instantiation of this class leads to undefined behavior, but it's needed
+   * for certain situations.
```
[the remark this answers is not in the document]
Comment (solatis) [reply quasardb/logger.hpp]: Actually after re-reading the code a few times, the lock in `swap_callback` actually ensures this isn't possible. But basically QuasarDB would need to start sending callbacks _before_ the _logger is initialized.

File: quasardb/logger.hpp (hunk 3 above)
Comment (solatis) [reply quasardb/logger.hpp]: Unfortunately, all these Python objects work with a reference to the 'Python Interpreter' under the hood. I don't think there is a strict guarantee that this reference will always be valid, only in the context of a function call.

File: quasardb/error.hpp (hunk 4 above)
Comment (solatis) [reply quasardb/error.hpp]: Good point.

File: quasardb/pandas/__init__.py (hunk 2 above)
Comment (solatis) [reply quasardb/pandas/__init__.py]: Yeah as this is somewhat user-visible, and relates to "Building a Pandas dataframe out of a number of Pandas series", I figured it was best to stick with Pandas terminology.

Comment (solatis) [reply quasardb/pandas/__init__.py]: But a DataFrame == a table, and a TimeSeries == a column per pandas <> qdb terminology, for what it's worth.

**Pull request merged**